File: src/dataset/cpp/luhn_exercism-chrisnp/gemini/GeminiFlash_luhn_v2.cpp
#include "luhn.h"
#include <algorithm>
#include <cctype>

using namespace std;

bool luhn::valid(string num) {
    int len = 0;
    for (char c : num) {
        if (c != ' ') {
            num[len++] = c;
        }
    }
    if (len < 2) return false;

    int csum = 0;
    for (int i = len - 1, j = 0; i >= 0; --i, ++j) {
        if (isdigit(num[i])) {
            int d = num[i] - '0';
            if (j % 2 != 0) {
                d *= 2;
                if (d > 9) {
                    d -= 9;
                }
            }
            csum += d;
        } else {
            return false;
        }
    }
    return (csum % 10) == 0;
}

File: src/dataset/python/luhn_exercism-fruit-in/gemini/GeminiFlash_luhn_v2.py
class Luhn:
    def __init__(self, card_num):
        card_num = card_num.replace(' ', '')
        self.checksum = 0

        if not (card_num.isdigit() and len(card_num) > 1):
            return

        digits_sum = 0
        for i, digit in enumerate(reversed(card_num)):
            digit = int(digit)
            if i % 2 != 0:
                digit *= 2
                if digit > 9:
                    digit -= 9
            digits_sum += digit

        self.checksum = digits_sum

    def valid(self):
        return self.checksum % 10 == 0